### PR TITLE
Only save a project if its creation is entirely successful

### LINKED ISSUE #1531: Keep any project creation from saving until completion in case of cancellation

### PR DESCRIPTION
**Pull request by chasefleming**
Closes #1531 

This creates the whole project in a system temporary directory until the entire project is successfully created, otherwise it is destroyed. Upon successful creation, it moves it out of the temp directory and into the target directory.
______

For contributor use:

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-cli/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

File: internal/super/setup.go [hunk 1]
```diff
@@ -77,134 +77,163 @@ func create(
...
+			logger.Error(fmt.Sprintf("Failed to remove %s: %v", tempDir, err))
 		}
+	}()
 
-		selectedContractNames, err := prompt.RunSelectOptions(contractNames, promptMessage)
-		if err != nil {
-			fmt.Printf("Error running dependency selection: %v\n", err)
-			os.Exit(1)
-		}
+	fmt.Printf("Creating project in %s\n", tempDir)
```
Comment (jribbink) [inline internal/super/setup.go:138]:
```suggestion
	fmt.Printf("Creating project in %s\n", targetDir)
```

File: internal/super/setup.go [hunk 2]
```diff
@@ -77,134 +77,163 @@ func create(
...
+				},
+			})
 		}
+	}
 
+	// Add the selected core contracts as dependencies
+	installer, err := dependencymanager.NewDependencyInstaller(logger, state, false, tempDir, dependencymanager.Flags{})
+	if err != nil {
+		logger.Error(fmt.Sprintf("Error: %v", err))
+		return "", err
```
Comment (jribbink) [inline internal/super/setup.go:211]:
```suggestion
		return "", err
```

this should be handled by returning an error from the command

File: internal/super/setup.go [hunk 3]
```diff
@@ -77,134 +77,163 @@ func create(
...
 	}
 
-	return &setupResult{targetDir: targetDir}, nil
+	if err := installer.AddMany(dependencies); err != nil {
+		logger.Error(fmt.Sprintf("Error: %v", err))
+		return "", err
```
Comment (jribbink) [inline internal/super/setup.go:216]:
```suggestion
		return "", err
```

File: internal/super/setup.go [hunk 4]
```diff
@@ -77,134 +77,163 @@ func create(
...
+	var err error
 
-		_, err = generateNew([]string{"DefaultContract"}, "contract", directoryPath, logger, state)
+	// Ask for project name if not given
+	if len(args) < 1 {
+		userInput, err := prompt.RunTextInput("Enter the name of your project", "Type your project name here...")
 		if err != nil {
-			return nil, err
+			fmt.Printf("Error running project name: %v\n", err)
+			os.Exit(1)
```
Comment (jribbink) [inline internal/super/setup.go:112]:
```suggestion
			return nil, fmt.Errorf("Error running project name: %v", err)
```

Is there a reason we can't use the standard error procedure?

File: internal/super/setup.go [hunk 5]
```diff
@@ -77,134 +77,163 @@ func create(
...
+	contractNames := make([]string, 0)
+
+	for _, contract := range sc.All() {
+		contractNames = append(contractNames, contract.Name)
+	}
+
+	selectedContractNames, err := prompt.RunSelectOptions(contractNames, promptMessage)
+	if err != nil {
+		fmt.Printf("Error running dependency selection: %v\n", err)
+		os.Exit(1)
```
Comment (jribbink) [inline internal/super/setup.go:188]:
```suggestion
		return "", fmt.Errorf("error running dependency selection: %v\n", err)
```

**Review by jribbink — COMMENTED** [review]
Awesome!  Just some nits about error handling

**Review by ianthpun — APPROVED** [review]

**Review by jribbink — APPROVED** [review]
LGTM, tested and works.

If we can just cleanup printing the temp paths that would be awesome:
image

**Pull request merged**